examples/chaincode/go/authorizable_counter/authorizable_counter.go

```diff
@@ -0,0 +1,122 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+
+	"github.com/hyperledger/fabric/core/chaincode/shim"
+)
+
+// AuthorizableCounterChaincode is an example that use Attribute Based Access Control to control the access to a counter by users with an specific role.
+// In this case only users which TCerts contains the attribute position with the value "Software Engineer" will be able to increment the counter.
+type AuthorizableCounterChaincode struct {
+}
+
+//Init the chaincode asigned the value "0" to the counter in the state.
+func (t *AuthorizableCounterChaincode) Init(stub *shim.ChaincodeStub, function string, args []string) ([]byte, error) {
+	err := stub.PutState("counter", []byte("0"))
+	return nil, err
+}
+
+//Invoke Transaction makes increment counter
+func (t *AuthorizableCounterChaincode) Invoke(stub *shim.ChaincodeStub, function string, args []string) ([]byte, error) {
+	isOk, _ := stub.VerifyAttribute("position", []byte("Software Engineer")) // Here the ABAC API is called to verify the attribute, just if the value is verified the counter will be incremented.
+	if isOk {
+		counter, err := stub.GetState("counter")
+		if err != nil {
+			return nil, err
+		}
+		var cInt int
+		cInt, err = strconv.Atoi(string(counter))
+		if err != nil {
+			return nil, err
+		}
+		cInt = cInt + 1
+		counter = []byte(strconv.Itoa(cInt))
+		stub.PutState("counter", counter)
+	}
+	return nil, nil
+
+}
+
+//Delete delete the counter from the state.
+func (t *AuthorizableCounterChaincode) delete(stub *shim.ChaincodeStub, args []string) ([]byte, error) {
+	if len(args) != 1 {
+		return nil, errors.New("Incorrect number of arguments. Expecting 3")
+	}
+
+	// Delete the key from the state in ledger
+	err := stub.DelState("counter")
+	if err != nil {
+		return nil, errors.New("Failed to delete state")
+	}
+
+	return nil, nil
+}
+
+// Run callback representing the invocation of a chaincode
+// This chaincode will increment the counter if the user has an attribute called "position" with the value "Software Engineer"
+func (t *AuthorizableCounterChaincode) Run(stub *shim.ChaincodeStub, function string, args []string) ([]byte, error) {
+
+	// Handle different functions
+	if function == "init" {
+		// Initialize the entities and their asset holdings
+		return t.Init(stub, function, args)
+	} else if function == "invoke" {
+		// Transaction makes payment of X units from A to B
+		return t.Invoke(stub, function, args)
+	} else if function == "delete" {
+		// Deletes an entity from its state
+		return t.delete(stub, args)
+	}
+
+	return nil, errors.New("Received unknown function invocation")
+}
+
+// Query callback representing the query of a chaincode
+func (t *AuthorizableCounterChaincode) Query(stub *shim.ChaincodeStub, function string, args []string) ([]byte, error) {
+	if function != "query" {
```

Comment:
Naming the function of the Query function "query" is confusing. It would be better to give it a different name.